### PR TITLE
Improve vault-manager performance

### PR DESCRIPTION
**Pull request by riuvshin**
Improve vault-manager performance

In this PR:
- stop using vault client directly by toplevel configuration blocks, so `pkg/vault/client.go` now exposes public methods to interact with Vault.
- all vault `read` operations now running in parallel to improve vault-manager performance
- Update logs
- Update tests

### CODE REVIEW

**Comment by maorfr** [comment]
i like the usage of wg a lot, but it means that everything in the same priority runs in parallel.
would it make sense to be able to control the number of parallel actions?
we can use a bounded weight group for that -
https://gist.github.com/maorfr/9f8ad44e732e489d498d6ddebf181a49

wdyt?

**Comment by riuvshin** [comment]
@maorfr that was exactly what I wanted to discuss with you :P

**Comment by maorfr** [comment]
i think that would make sense, and it will also comply with qontract-reconcile (we added a `--thread-pool-size` optional flag with a default value of 10).

**Comment by riuvshin** [comment]
yeah something like that! because imagine we have 10000 policies that will run 10000 routines :O

**Comment by maorfr** [comment]
exactly. so bwg it is!

**Comment by riuvshin** [comment]
@maorfr `bwg` is great! I've updated PR can you please review it again :)

File: toplevel/audit/audit.go [hunk 1]
```diff
@@ -74,50 +57,70 @@ func init() {
...
-		log.WithField("package", "audit").WithError(err).Fatal("failed to decode audit device configuration")
+		log.WithError(err).Fatal("[Vault Audit] failed to decode audit device configuration")
 	}
 
-	// Get the existing enabled Audits Devices.
-	enabledAudits, err := vault.Client().Sys().ListAudit()
-	if err != nil {
-		log.WithField("package", "audit").WithError(err).Fatal("failed to list audit devices from Vault instance")
-	}
+	enabledAudits := vault.ListAuditDevices()
```
Comment (maorfr) [inline toplevel/audit/audit.go]: yay!

File: toplevel/auth/auth.go [hunk 2]
```diff
@@ -143,16 +137,36 @@ func (c config) Apply(entriesBytes []byte, dryRun bool) {
...
+
+						mutex.Lock()
+
+						policyMappingPath := filepath.Join("/auth/", e.Path, "map/teams", entities[i].(string))
+						policiesMappedToEntity := vault.ReadSecret(policyMappingPath).Data["value"].(string)
+						policies := make([]map[string]interface{}, 0)
+						for _, policy := range strings.Split(policiesMappedToEntity, ",") {
+							policies = append(policies, map[string]interface{}{"name": policy})
+						}
+						existingPolicyMappings = append(existingPolicyMappings,
```
Comment (maorfr) [inline toplevel/auth/auth.go]: I would lock just before this line, as this is the only one accessing a shared resource (one that is defined outside the scope of the go routine)

File: toplevel/policy/policy.go [hunk 3]
```diff
@@ -37,52 +39,63 @@ func (e entry) Equals(i interface{}) bool {
...
+
+			bwg.Add(1)
+
+			go func(i int) {
+
+				mutex.Lock()
+
+				name := existingPolicyNames[i]
+				policy := vault.GetVaultPolicy(name)
+				existingPolicies = append(existingPolicies, entry{Name: name, Rules: policy})
```
Comment (maorfr) [inline toplevel/policy/policy.go]: same thing regarding lock

File: toplevel/role/role.go [hunk 4]
```diff
@@ -88,18 +85,35 @@ func (c config) Apply(entriesBytes []byte, dryRun bool) {
...
+				for i := range roles {
+
+					bwg.Add(1)
+
+					go func(i int) {
+
+						mutex.Lock()
+
+						path := filepath.Join("auth", authBackend, "role", roles[i].(string))
+						existingRoles = append(existingRoles, entry{
```
Comment (maorfr) [inline toplevel/role/role.go]: same thing regarding lock

File: toplevel/auth/auth.go (hunk 2 above)
Comment (riuvshin) [reply toplevel/auth/auth.go]: totally!

File: toplevel/policy/policy.go (hunk 3 above)
Comment (riuvshin) [reply toplevel/policy/policy.go]: +1000

File: toplevel/role/role.go (hunk 4 above)
Comment (riuvshin) [reply toplevel/role/role.go]: +1000

File: toplevel/auth/auth.go (hunk 2 above)
Comment (riuvshin) [reply toplevel/auth/auth.go]: changes pushed!

File: toplevel/policy/policy.go (hunk 3 above)
Comment (riuvshin) [reply toplevel/policy/policy.go]: changes pushed!

File: toplevel/role/role.go (hunk 4 above)
Comment (riuvshin) [reply toplevel/role/role.go]: changes pushed!

**Pull request merged**